creator/releases/tasks.py

```diff
@@ -0,0 +1,574 @@
+            kf_id=service["kfId"], defaults=defaults
+        )
+        try:
+            release_service.creator = User.objects.get(
+                username=service["author"]
+            )
+        except User.DoesNotExist:
+            pass
+
+        defaults = {
+            "uuid": task["uuid"],
+            "created_at": task["createdAt"],
+            "state": task["state"],
+            "release": release,
+            "release_service": release_service,
+        }
+        release_task, created = ReleaseTask.objects.update_or_create(
+            kf_id=task["kfId"], defaults=defaults
+        )
+        tasks.append(release_task)
+
+    logger.info(
+        f"Synced new release '{release.kf_id}' with "
+        f"{len(studies)} studies and {len(tasks)} tasks."
+    )
+
+
+@task("release")
+def initialize_release(release_id=None):
+    """
+    Initializes a release by queueing up jobs to initialize each task in the
+    release.
+    """
+    release = Release.objects.get(pk=release_id)
+
+    if release.tasks.count() == 0:
+        logger.info(
+            f"No services were requested to run for release '{release.pk}'. "
+            f"The release will be automatically moved to the 'running' state."
+        )
+        release.start()
+        release.save()
+
+        queue = django_rq.get_queue("releases")
+        queue.enqueue(
+            start_release, release_id=release.pk, ttl=settings.RQ_DEFAULT_TTL
+        )
+        return
+
+    logger.info(
+        f"Queuing jobs to check that {release.tasks.count()} tasks are "
+        f"prepared to start processing release '{release.pk}'"
+    )
+
+    for task in release.tasks.all():
+        logger.info(
+            f"Queuing initialize_task for task '{task.pk}' "
+            f"of service '{task.release_service.name}'"
+        )
+        queue = django_rq.get_queue("releases")
+        queue.enqueue(
+            initialize_task, task_id=task.pk, ttl=settings.RQ_DEFAULT_TTL
+        )
+
+
+@task("release_task")
+def initialize_task(task_id=None):
+    """
+    Initializes a task by sending it the initialize command.
+    """
+    task = ReleaseTask.objects.get(pk=task_id)
+    try:
+        task.initialize()
+        task.save()
+    except Exception as err:
+        logger.error(
+            f"There was a problem trying to initialize the service: {err}"
+        )
+        logger.info(
+            "The release will be canceled as the service was not ready to "
+            "start a new release"
+        )
+        task.reject()
+        task.save()
+
+        task.release.cancel()
+        task.release.save()
+
+        queue = django_rq.get_queue("releases")
+        queue.enqueue(
+            cancel_release,
+            release_id=task.release.pk,
+            failed=True,
+            ttl=settings.RQ_DEFAULT_TTL,
+        )
+
+    # Check if all tasks are initialized now and queue up the release start
```

Comment:
I'm wondering why this block is here instead of at the end of the `initialize_release` task? It looks like the `start_release` task will be queued for each task when it should be queued once for just the release?

Reply:
Still wondering :)

Reply:
Ok I see what you're doing now. My second thought was could `initialize_task` just be a method that gets called in the task loop on line 201, but that would make the initializing all of the tasks a synchronous process which I presume you don't want